Misc improvement: `suspendmanager.lua` has new suspension reason for unsupported constructions.

master-spike opened the pull request:
This change to suspendmanager deals with https://github.com/DFHack/dfhack/issues/3601. I've added a new suspension reason `UNSUPPORTED` and added some checks relating to it. The way it works is based on an observation: each tiletype shape that can provide support provides it in one of two ways - either it supports like a wall (all 6 ortho axes as well as orthogonally on the layer above, with an invisible floor), or it supports like a floor (orthogonally on the same level). The only other thing that can provide support is a support building (pillar), which provides support to floors above it or walls both above and below.

I have tested this to a standard that I hope is rigorous enough although @plule says he's willing to take a shot at testing some other cases I may have missed.

donhth commented on suspendmanager.lua:
```diff
@@ -266,6 +340,134 @@ local function neighbours(pos)
+    local constr_type = building:getSubtype()
+    if CONSTRUCTION_FLOOR_SUPPORT[constr_type] then
+        for _,n in pairs(neighboursWallSupportsFloor(pos)) do
+            local tt = dfhack.maps.getTileType(n)
+            if tt then
+                local attrs = df.tiletype.attrs[tt]
+                if TILETYPE_SHAPE_WALL_SUPPORT[attrs.shape] then  return false end
+            end
```
You might consider breaking these bits out into a function.

plule commented:
Maybe I'm mixing up something but it does not seem to correctly detect this situation, buidling a gold bar floor:

![image](https://github.com/DFHack/scripts/assets/630159/d8ab4e05-597c-4e0d-bc78-7d68d0b31168)

below:

![image](https://github.com/DFHack/scripts/assets/630159/92ede901-bef4-4440-9dd2-473f4692cce9)

above:

![image](https://github.com/DFHack/scripts/assets/630159/4cd5e11c-2b89-4f40-9885-90332d0f67f6)

master-spike commented on suspendmanager.lua:
```diff
@@ -161,6 +164,77 @@ local CONSTRUCTION_IMPASSABLE = utils.invert{
+}
+
+local CONSTRUCTION_FLOOR_SUPPORT = utils.invert{
+    df.construction_type.FLOOR,
```
Good catch, that must be why it doesn't work for floors!

master-spike replied:
I've made the change now - should work for that case

plule replied:
It does :)

plule commented on suspendmanager.lua:
```diff
@@ -433,6 +639,11 @@ function SuspendManager:refresh()
             end
         end
 
+        -- Check for construction jobs which may be unsupported
```
This block should be moved under the ```Internal reasons to suspend a job```.

The terminology is not very good, but everything before this check is what is already suspended by dwarf fortress, and suspendmanager (/unsuspend) will never unsuspend it, and everything after is what suspendmanager is actively suspending.

master-spike replied:
Ok, I misunderstood what `preventBlocking` meant in this context. I think maybe that might warrant a rename due to the expanded scope of this script? perhaps something like `preventRiskyJobs` or something?

plule replied:
It's actually even wider than that, the suspended jobs on smoothing designation also fall in that category, they are not risky, they would just erase a smoothing job. Maybe falling back on the broader "smart" term would be best.

This issue is spread all across the suspension script, even in the UI and the `suspend`/`unsuspend` command line options, I suggest not to do that in that pr, maybe a separate issue?

plule commented on suspendmanager.lua:
```diff
@@ -36,6 +36,7 @@ REASON = {
     ERASE_DESIGNATION = 4,
     --- Blocks a dead end (either a corridor or on top of a wall)
     DEADEND = 5,
+    UNSUPPORTED = 6,
```
```suggestion
    --- Would cave-in immediately on completion
    UNSUPPORTED = 6,
```

plule commented:
With floors and tracks (at least), it won't suspend it if there is a floor above, but it does cave-in:

![image](https://github.com/DFHack/scripts/assets/630159/fe723d20-14b0-44f8-b31a-b15a37a6a10b)

above:
![image](https://github.com/DFHack/scripts/assets/630159/2cd3543e-7d59-40cb-8da5-9800b635dfb1)

below:
![image](https://github.com/DFHack/scripts/assets/630159/9ee37761-1e5a-4f10-9814-32a66dc752e3)

master-spike commented:
> With floors and tracks (at least), it won't suspend it if there is a floor above, but it does cave-in:
> 
> ![image](https://user-images.githubusercontent.com/630159/270408216-fe723d20-14b0-44f8-b31a-b15a37a6a10b.png)
> 
> above: ![image](https://user-images.githubusercontent.com/630159/270408279-2cd3543e-7d59-40cb-8da5-9800b635dfb1.png)
> 
> below: ![image](https://user-images.githubusercontent.com/630159/270408336-9ee37761-1e5a-4f10-9814-32a66dc752e3.png)

I think I made a mistake copying and pasting the coordinates and forgetting to change them 🤦 I think I've fixed this and I've added comments on the directions to make it easier to quickly verify correctness

plule commented on suspendmanager.lua:
```diff
@@ -266,6 +341,138 @@ local function neighbours(pos)
+    if not building or building:getType() ~= df.building_type.Construction then return false end
+
+    local pos = {x=building.centerx, y=building.centery,z=building.z}
+
```
One small quality of life thing: It feels a bit "spammy" to suspend unreachable constructions:

![image](https://github.com/DFHack/scripts/assets/630159/ed5c79c9-9775-4b9a-9e6a-ac840b68ea9d)

Maybe you could add an early return here if none of the direct neighbour is walkable to make suspendmanager less intrusive in that situation.

plule replied:
mmh this broke it actually, walkable() returns false for floor bars, the tile type shape is «empty» for walkable constructions.

I'm trying to understand the consequences for suspendmanager in general, this could cause other bugs

plule replied:
@master-spike My brain is too fried to understand/find the consequences of the `walkable()` bug, maybe you can revert/comment out this part for now.

As far as I'm concerned except this, it's all good :)

master-spike replied:
Ah, it's because `walkable()` is incomplete - we don't have definitions of walkable buildings (floor bars/grates, retracted bridges, etc) like we do for impassable ones. I'll roll back this change and maybe see if there's a better way to check for walkable tiles.

master-spike replied:
Probably better to fix `walkable()` in a separate issue

plule commented:
For the terminology, I think that "would collapse" is the best for any user facing text (so reason description «Would collapse immediately», doc and changelog too?), it's the term used in the native UI «Something has collapsed on the surface!». «Unsupported» feels a bit ambiguous, and cave in is not that obvious to me. Disclaimer I'm not a native speaker though

master-spike commented:
> For the terminology, I think that "would collapse" is the best for any user facing text (so reason description «Would collapse immediately», doc and changelog too?), it's the term used in the native UI «Something has collapsed on the surface!». «Unsupported» feels a bit ambiguous, and cave in is not that obvious to me. Disclaimer I'm not a native speaker though

You're right, something about this was bothering me too. "Unsupported" sounds like it could be mistaken for a technical term (as in, it could be read like "this script doesn't support this type of construction"), rather than a game thing.

donhth commented:
You can also say it lacks a foundation, has no foundation, needs foundation, or requires a foundation, etc.

plule reviewed (approved):
Nice!
